### PR TITLE
fix: oauth client metadata as object in CRD (#71)

### LINKED ISSUE #71: OAuth2 client CRD metadata issue
Hi folks,

Any reason why in recent release metadata type in `hydra.ory.sh_oauth2clients.yaml` has changed from object to byte string?

**Expected**:

```
   metadata:
      description: Metadata is abritrary data
      type: object
```

**Actual**:
```
   metadata:
      description: Metadata is abritrary data
      format: byte
      type: string
```
The only way I found to set object there is to `base64encode` the json string. But, in our case it breaks IDP, as we should do base64 decoding there. 
And, it also seems that Hydra `/clients` endpoint didn't change in that perspective - metadata is still an object.

Previously, we were able to specify it in our `yaml` like this:
```
metadata:
    type: internal

```
And get response from hydra with appropriate object:
```
 {
   ...
   metadata: {
       'type': 'internal'
   }

```

And now if we want to specify object, we specify base64 encoded json string in yaml:
```
metadata: eyJ0eXBlIjoiaW50ZXJuYWwifQ==

```
And we get it as-is from hydra `/clients` endpoints:
```
 {
   ...
   metadata: "eyJ0eXBlIjoiaW50ZXJuYWwifQ=="
 }
```

Please, let me know if it was done purposely or is maybe some auto-generation issue?
I can do the PR, just let me know if I'm missing some puzzle here.

@colunira @aeneasr @Demonsthere

### PR DESCRIPTION
**Pull request by romanlytvyn**
Fixes #71 

Simply adding `+kubebuilder:validation:Type=object` annotation is not enough, as it results in `conflicting types in allOf branches in schema: string vs object` error.

Switched `Metadata` type from `json.RawMessage` to `JSON` from `apiextensions` package, which represents any valid JSON value.
These types are supported: bool, int64, float64, string, []interface{}, map[string]interface{} and nil.
https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1#JSON

### CODE REVIEW

**Review by Demonsthere — CHANGES_REQUESTED** [review]
Hello there! Thanks for the PR ;) Just one thing I would change

File: api/v1alpha1/oauth2client_types.go [hunk 1]
```diff
@@ -182,6 +187,8 @@ func init() {
 
 // ToOAuth2ClientJSON converts an OAuth2Client into a OAuth2ClientJSON object that represents an OAuth2 client digestible by ORY Hydra
 func (c *OAuth2Client) ToOAuth2ClientJSON() *hydra.OAuth2ClientJSON {
+	meta, _ := json.Marshal(c.Spec.Metadata)
```
[the remark this answers is not in the document]
Comment (romanlytvyn) [reply api/v1alpha1/oauth2client_types.go]: good point, added json encoding error propagation

File: controllers/oauth2client_controller.go [hunk 2]
```diff
@@ -205,16 +205,24 @@ func (r *OAuth2ClientReconciler) registerOAuth2Client(ctx context.Context, c *hy
 		return err
 	}
 
+	oauth2client, err := c.ToOAuth2ClientJSON()
+	if err != nil {
+		if updateErr := r.updateReconciliationStatusError(ctx, c, hydrav1alpha1.StatusRegistrationFailed, err); updateErr != nil {
+			return updateErr
+		}
+		return nil
```
Comment (Demonsthere) [inline controllers/oauth2client_controller.go:213]: Shouldn't we return the original error here?

File: controllers/oauth2client_controller.go [hunk 3]
```diff
@@ -257,7 +265,15 @@ func (r *OAuth2ClientReconciler) updateRegisteredOAuth2Client(ctx context.Contex
 		return err
 	}
 
-	if _, err := hydra.PutOAuth2Client(c.ToOAuth2ClientJSON().WithCredentials(credentials)); err != nil {
+	oauth2client, err := c.ToOAuth2ClientJSON()
+	if err != nil {
+		if updateErr := r.updateReconciliationStatusError(ctx, c, hydrav1alpha1.StatusUpdateFailed, err); updateErr != nil {
+			return updateErr
+		}
+		return nil
```
Comment (Demonsthere) [inline controllers/oauth2client_controller.go:273]: Also here

File: controllers/oauth2client_controller.go [hunk 4]
```diff
@@ -205,16 +205,24 @@ func (r *OAuth2ClientReconciler) registerOAuth2Client(ctx context.Context, c *hy
 		return err
 	}
 
+	oauth2client, err := c.ToOAuth2ClientJSON()
+	if err != nil {
+		if updateErr := r.updateReconciliationStatusError(ctx, c, hydrav1alpha1.StatusRegistrationFailed, err); updateErr != nil {
+			return updateErr
+		}
+		return err
```
Comment (Demonsthere) [inline controllers/oauth2client_controller.go:213]:
```suggestion
		return errors.WithStack(err)
```

File: controllers/oauth2client_controller.go [hunk 5]
```diff
@@ -257,7 +265,15 @@ func (r *OAuth2ClientReconciler) updateRegisteredOAuth2Client(ctx context.Contex
 		return err
 	}
 
-	if _, err := hydra.PutOAuth2Client(c.ToOAuth2ClientJSON().WithCredentials(credentials)); err != nil {
+	oauth2client, err := c.ToOAuth2ClientJSON()
+	if err != nil {
+		if updateErr := r.updateReconciliationStatusError(ctx, c, hydrav1alpha1.StatusUpdateFailed, err); updateErr != nil {
+			return updateErr
+		}
+		return err
```
Comment (Demonsthere) [inline controllers/oauth2client_controller.go:273]:
```suggestion
		return errors.WithStack(err)
```

**Review by Demonsthere — COMMENTED** [review]
Looking nice :)

File: controllers/oauth2client_controller.go (hunk 2 above)
Comment (romanlytvyn) [reply controllers/oauth2client_controller.go]: fixed

File: controllers/oauth2client_controller.go (hunk 3 above)
Comment (romanlytvyn) [reply controllers/oauth2client_controller.go]: fixed

File: controllers/oauth2client_controller.go (hunk 4 above)
Comment (romanlytvyn) [reply controllers/oauth2client_controller.go]: added

File: controllers/oauth2client_controller.go (hunk 5 above)
Comment (romanlytvyn) [reply controllers/oauth2client_controller.go]: added :)

**Review by Demonsthere — APPROVED** [review]
LGTM :)

**Pull request merged**